sql/expression/function/aggregation/window_functions.go

```diff
@@ -1211,6 +1211,150 @@ func (a *PercentRank) Compute(ctx *sql.Context, interval sql.WindowInterval, buf
 	}
 }
 
+type Rank struct {
```

Comment:
i'd collapse the 3 rank functions into a `rankBase` if possible, percent rank doesn't need extra fields but you could use an enum field to differentiate the 3, dense rank will require the two extra vars. The compute function will switch on the rank type. Alternatively, if the compute function for the 3 combined becomes really gnarly, you could instead have `Rank`, `PercentRank`, and `DenseRank` overload `Compute` methods, with `DenseRank` and `PercentRank` tracking their own corresponding states relative to the child `baseRank`.